public/app/features/trails/MetricSelectScene.tsx

```diff
@@ -90,6 +91,11 @@ export class MetricSelectScene extends SceneObjectBase<MetricSelectSceneState> {
       // Temp hack when going back to select metric scene and variable updates
       this.ignoreNextUpdate = true;
     }
+
+    const dataSourceVariable = sceneGraph.lookupVariable(VAR_DATASOURCE, this);
+    dataSourceVariable?.subscribeToState((state) => {
+      this.onDataSourceChange();
+    });
```

Comment:
Try something like this instead: 

 ```ts
protected _variableDependency = new VariableDependencyConfig(this, {
    variableNames: [VAR_METRIC_NAMES, VAR_DATASOURCE],
    onReferencedVariableValueChanged: this.onVariableChanged.bind(this),
  });

  private onVariableChanged(variable: SceneVariable): void {
    if (variable.state.name === VAR_DATASOURCE) {
      // Clear filtered metrics
      this.setState({ metricsAfterFilter: undefined, metricsAfterSearch: undefined });
      // Clear all panels for the previous data source
      this.state.body.setState({ children: [] });
    }

    if (variable.state.name === VAR_METRIC_NAMES) {
      this.updateMetrics(); // Entire pipeline must be performed
      this.buildLayout();
    }
  }
```

this would remove onVariableUpdateCompleted, we only care about value changes in this component so using onReferencedVariableValueChanged is better.

Oh and let scenes lib worry about cancelling the metric names variable

Reply:
https://github.com/grafana/scenes/pull/557

Reply:
I was wondering:
Any preference to an instance method `onVariableChanged` vs. a local nameless lambda?